File: src/chapters/nonlinear-equations/content/code/fixed-point.r
fixed_point <- function(g, x0, tol = 1e-12, max_iter = 200) {
  # Fixed-point iteration x_{k+1} = g(x_k).
  x <- x0
  for (k in 1:max_iter) {
    xn <- g(x)
    if (abs(xn - x) < tol) return(list(root = xn, iters = k))
    x <- xn
  }
  list(root = x, iters = max_iter)
}

res <- fixed_point(cos, 1.0)        # Dottie number ~0.739085
cat(sprintf("(%.16g, %d)\n", res$root, res$iters))

File: src/chapters/nonlinear-equations/content/code/fixed-point.rs
// Fixed-point iteration x_{k+1} = g(x_k).
fn fixed_point<G: Fn(f64) -> f64>(g: G, x0: f64, tol: f64, max_iter: usize) -> f64 {
    let mut x = x0;
    for _ in 0..max_iter {
        let xn = g(x);
        if (xn - x).abs() < tol { return xn; }
        x = xn;
    }
    x
}
fn main() {
    println!("{}", fixed_point(|x: f64| x.cos(), 1.0, 1e-12, 200));
}
